library/alloc/src/vec/source_iter_marker.rs

```diff
@@ -101,6 +101,8 @@ fn write_in_place_with_drop<T>(
 trait SpecInPlaceCollect<T, I>: Iterator<Item = T> {
     /// Collects an iterator (`self`) into the destination buffer (`dst`) and returns the number of items
     /// collected. `end` is the last writable element of the allocation and used for bounds checks.
+    // FIXME: Clarify safety conditions. Iterator must not be coerced to a subtype
+    // after this call due to potential use of [`TrustedRandomAccessNoCoerce`].
```

Comment:
```suggestion
    ///
    /// # Safety
    /// 
    /// This method is specialized and one of its implementations makes use of the
    /// `TrustedRandomAccessNoCoerce` contract which means the caller must also
    /// uphold that contract. It could be trivially upheld by making no further use of the iterator and dropping it.
    /// But to actually achieve in-place iteration we must also forget the source allocation
    /// in `vec::IntoIter` after `get_unchecked` has been used by the specialization.
    /// This should interact safely with `TrustedRandomAccessNoCoerce`
    /// because the `SourceIter::as_inner` requires that adapters' `Drop` impls must not rely on
    /// the underlying source remaining the same between iteration and the drop call.
    /// Additionally `TrustedRandomAccessNoCoerce` is only implemented for `Copy`
    /// types which ensures that `vec::IntoIter` does not have to perform any drops for its contents.
```

Reply:
I don’t quite understand the relevance of “`TrustedRandomAccessNoCoerce` is only implemented for `Copy`”. This doesn’t seem like it’s guaranteed anywhere and it also doesn’t seem relevant for the use case with `Vec::IntoIter` and its `forget_allocation_drop_remaining` method.

Reply:
Also, this is not technically (marked as) an `unsafe` fn, so full-blown `# Safety` comment feels possibly overkill. I guess since it’s a private trait only used locally in the `from_iter` implementaiton of
```rust
impl<T, I> SpecFromIter<T, I> for Vec<T>
where
    I: Iterator<Item = T> + SourceIterMarker,
 ```
 so it might be okay? I’d be fine with a note here that just adds visibility to the fact that `__iterator_get_unchecked` is used internally so that nobody forgets about the requirements that that might entail.

Reply:
So, in terms of

> a note here that just adds visibility to the fact that `__iterator_get_unchecked` is used internally so that nobody forgets about the requirements that that might entail

What would you think of something like
```rust
/// Helper trait to hold specialized implementations of the in-place iterate-collect loop
trait SpecInPlaceCollect<T, I>: Iterator<Item = T> {
    /// Collects an iterator (`self`) into the destination buffer (`dst`) and returns the number of items
    /// collected. `end` is the last writable element of the allocation and used for bounds checks.
    ///
    /// This method is specialized and one of its implementations makes use of
    /// `Iterator::__iterator_get_unchecked` calls with a `TrustedRandomAccessNoCoerce` bound
    /// on `I` which means the caller of this method must take the safety conditions
    /// of that trait into consideration.
    fn collect_in_place(&mut self, dst: *mut T, end: *const T) -> usize;
}
```
for now? @the8472

Reply:
>  This doesn’t seem like it’s guaranteed anywhere

Well, this is a private trait used for in-place collection, and we can only collect in place on `vec::IntoIter` right now which only implements TRANC for T: Copy.

> also doesn’t seem relevant for the use case with Vec::IntoIter and its forget_allocation_drop_remaining method.

`T: Copy` means it doesn't have to drop. Which means it doesn't need know how remains to be dropped, because the amount of `remaining` items is unknown when trusted access is used since it doesn't update the pointers. It would try to drop the initial range of items before iteration rather the range after iteration which obviously would be wrong.


This stuff requires some whole-library knowledge to explain.

Reply:
> What would you think of something like [...]

Yeah, I guess as warning to look deeper it's enough.

The more detailed explanation should probably go at the callsite or the module level instead, even though there already are plenty of comments they may not be the right ones since it has been mostly me so far who worked on this code.

Reply:
> _[…]_
>
> This stuff requires some whole-library knowledge to explain.

Okay, I do now understand what you’re talking about. I’m absolutely not sure where the right place to document this stuff would be.

Reply:
I’ve written a longer `FIXME` on the `forget_allocation_drop_remaining` call right now. I can commit this together with the commend presented in my previous post.

```rust
        // drop any remaining values at the tail of the source
        // but prevent drop of the allocation itself once IntoIter goes out of scope
        // if the drop panics then we also leak any elements collected into dst_buf
        //
        // FIXME: Since `SpecInPlaceCollect::collect_in_place` above might use
        // `__iterator_get_unchecked` internally, this call might be operating on
        // a `vec::IntoIter` with incorrect internal state regarding which elements
        // have already been “consumed”. However, the `TrustedRandomIteratorNoCoerce`
        // implementation of `vec::IntoIter` is only present if the `Vec` elements
        // don’t have a destructor, so it doesn’t matter if elements are “dropped multiple times”
        // in this case.
        // This argument technically currently lacks justification from the `# Safety` docs for
        // `SourceIter`/`InPlaceIterable` and/or `TrustedRandomAccess`, so it might be possible that
        // someone could inadvertently create new library unsoundness
        // involving this `.forget_allocation_drop_remaining()` call.
        src.forget_allocation_drop_remaining();
 ```
 
 Would this be enough for now or do we need to resolve these (internal) documentation issues before landing the PR? What do you think?

Reply:
I think the `FIXME` is quite long and would raise some eyebrows during review.

At this point it would be easier to just put a note on the `TrustedRandomIteratorNoCoerce` impl for `IntoIter` and a short one on the callsite. It's not really a property of the unsafe traits but of the vec-specific implementation. One Vec part relies on another Vec part not doing certain things to be safe.

Reply:
Well… it’s not just about `Vec`, it’s also kind-of about the whole story of accessing the `Vec` through an arbitrarily nested chain of iterator adaptors that preserve `InPlaceIteratble` and `TrustedRandomAccess`. Feel free to propose some concrete documentation, if you feel like it’s really necessary right now. Also note that this is somewhat related to the other `FIXME` in the `TrustedRandomAccess` docs stating
```rust
// FIXME: Clarify interaction with SourceIter/InPlaceIterable. Calling `SouceIter::as_inner`
// after `__iterator_get_unchecked` is supposed to be allowed.
```

I’m off for today, you can find the suggested changes from my comments so far here: https://github.com/steffahn/rust/commit/18a6d4541966042220c12b32d8c565bc415dbc9e.

By the way, if we want to restructure `TrustedRandomAccess` anyway, perhaps it doesn’t even make all that much sense to spend a lot of time fully documenting every aspect of the safety contract and interactions with `InPlaceIterable` etc. and writing it down nicely.

<hr>

I also don’t necessarily feel like such a long “FIXME” comment necessarily would lead to problems during further review. Anyone working with these traits or improving the situation in the future will be happy about a more lengthy explanation for this subtle point.

Maybe the `FIXME`s should somehow more clearly indicate that it’s _just_ about **fixing the _std-internal_ documentation**, and not about fixing the implementation.

Reply:
> if you feel like it’s really necessary right now.

Since this isn't new code I think it makes sense to find the right place and wording in a separate PR.

Unless reviewers need to retrace the existing logic, then such comments might help them.